test/integration/gardenlet/workloadidentity/tokenrequestor/tokenrequestor_test.go

```diff
@@ -0,0 +1,111 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	securityv1alpha1 "github.com/gardener/gardener/pkg/apis/security/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+)
+
+var _ = Describe("WorkloadIdentity TokenRequestor tests", func() {
+	var (
+		resourceName string
+
+		secret           *corev1.Secret
+		workloadIdentity *securityv1alpha1.WorkloadIdentity
+	)
+
+	BeforeEach(func() {
+		resourceName = "test-" + utils.ComputeSHA256Hex([]byte(uuid.NewUUID()))[:8]
+
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resourceName,
+				Namespace: testNamespace.Name,
+				Annotations: map[string]string{
+					"workloadidentity.security.gardener.cloud/name":      resourceName,
+					"workloadidentity.security.gardener.cloud/namespace": testNamespace.Name,
+				},
+				Labels: map[string]string{
+					"security.gardener.cloud/purpose": "workload-identity-token-requestor",
+				},
+			},
+		}
+
+		workloadIdentity = &securityv1alpha1.WorkloadIdentity{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resourceName,
+				Namespace: testNamespace.Name,
+				Labels: map[string]string{
+					testID: testRunID,
+				},
+			},
+			Spec: securityv1alpha1.WorkloadIdentitySpec{
+				Audiences: []string{"foo"},
+				TargetSystem: securityv1alpha1.TargetSystem{
+					Type: "foocloud",
+				},
+			},
+		}
+
+		fakeClock.SetTime(time.Now().Round(time.Second))
+	})
+
+	It("should behave correctly when: create w/o label, update w/ label, delete w/ label", func() {
+		secret.Labels = nil
+		Expect(testClient.Create(ctx, secret)).To(Succeed())
+		Expect(testClient.Create(ctx, workloadIdentity)).To(Succeed())
+		Expect(testClient.Get(ctx, client.ObjectKeyFromObject(workloadIdentity), workloadIdentity)).To(Succeed())
+
+		Consistently(func(g Gomega) {
+			g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+			g.Expect(secret.Data["token"]).To(BeEmpty())
+		}).Should(Succeed())
+
+		secret.Labels = map[string]string{"security.gardener.cloud/purpose": "workload-identity-token-requestor"}
+		Expect(testClient.Update(ctx, secret)).To(Succeed())
+
+		Eventually(func(g Gomega) {
+			g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+			g.Expect(secret.Data["token"]).ToNot(BeEmpty())
+
+			token, err := jwt.ParseSigned(string(secret.Data["token"]), []jose.SignatureAlgorithm{jose.ES256})
+			Expect(err).ToNot(HaveOccurred())
+
+			claims := jwt.Claims{}
+			Expect(token.Claims(verificationKey, &claims)).To(Succeed())
+			Expect(claims.Issuer).To(Equal("https://local.gardener.cloud"))
+			Expect(claims.Audience).To(Equal(jwt.Audience{"foo"}))
+			Expect(claims.Subject).To(Equal(fmt.Sprintf("gardener.cloud:workloadidentity:%s:%s:%s", testNamespace.Name, resourceName, workloadIdentity.UID)))
+		}).Should(Succeed())
+
+		Expect(testClient.Delete(ctx, secret)).To(Succeed())
+	})
+
+	AfterEach(func() {
```

Comment:
Shouldn't this be put into a `DeferCleanup` in the `BeforeEach`?

Reply:
Yes, this can also work.